src/annotator/anchoring/pdf.js

```diff
@@ -360,17 +398,59 @@ async function anchorQuote(quoteSelector, positionHint) {
       }
     }
 
-    const match = matchQuote(text, quoteSelector.exact, {
-      prefix: quoteSelector.prefix,
-      suffix: quoteSelector.suffix,
-      hint,
+    const [strippedText, offsets] = stripSpaces(text);
+
+    // Convert expected offset in original text into offset into stripped text.
+    let strippedHint;
+    if (hint !== undefined) {
+      strippedHint = 0;
+      while (strippedHint < offsets.length && offsets[strippedHint] < hint) {
+        ++strippedHint;
+      }
+    }
+
+    const match = matchQuote(strippedText, strippedQuote, {
+      prefix: strippedPrefix,
+      suffix: strippedSuffix,
+      hint: strippedHint,
     });
+
     if (!match) {
-      // We currently stop searching when we find a "good enough" match. This
-      // may not be the best match however. See https://github.com/hypothesis/client/issues/3705.
       continue;
     }
 
+    if (!bestMatch || match.score > bestMatch.match.score) {
+      bestMatch = {
+        page,
+        match: {
+          start: offsets[match.start],
+
+          // `match.end` is the offset one past the last character of the match
+          // in the stripped text. We need the offset one past the corresponding
+          // character in the original text.
+          //
+          // We assume here that matches returned by `matchQuote` must have
+          // be non-empty, so `match.end` > `match.start`.
+          end: offsets[match.end - 1] + 1,
+
+          score: match.score,
+        },
+      };
+
+      // If we found an exact match, stop early. This improves search performance
+      // in long documents compared to testing every page.
+      //
+      // A known limitation here is that the quote context (prefix and suffix)
+      // is not considered.
+      if (strippedText.slice(match.start, match.end) === strippedQuote) {
```

Comment:
Why not to fix this limitation?

It could be  something like this, I believe:

```
      // If we found an exact match, stop early. This improves search performance
      // in long documents compared to testing every page.
      const isExactQuote =
        strippedText.slice(match.start, match.end) === strippedQuote;
      const isExactPrefix =
        strippedPrefix === undefined
          ? true
          : strippedText.slice(
              match.start - strippedPrefix.length,
              match.start
            ) === strippedPrefix;
      const isExactSuffix =
        strippedSuffix === undefined
          ? true
          : strippedText.slice(
              match.end,
              match.end + strippedSuffix.length + 1
            ) === strippedPrefix;
      if (isExactQuote && isExactPrefix && isExactSuffix) {
```

Reply:
Yes, I think this is a decent trade-off between keeping the heuristics simple, finding the best match and optimizing the search. I've added a commit (baebd97a72f636854b26ffdb7de1d9e54b13d22e) which applies the basic idea from this proposal but handles missing `prefix` and `suffix` fields slightly:

- If the quote selector has either field, then require a match in at least one of them
- If the quote selector does not have either field, then don't require a match

The code as written in the proposal will ignore the `prefix` if the selector has a suffix and vice-versa.

Reply:
👍🏼